Coupling functor check

lindsayad opened the pull request:
This adds a new action that checks to see whether we should be adding
intra-element dof-coupling by inspecting the kernel and
boundary condition warehouses. If we do need intra-dof coupling
then we add a default coupling functor and reinit the
libMesh::ImplicitSystem.

lindsayad committed:
Don't add ghosting to mesh if not GEOMETRIC

I think our general theme is to require that users be explicit
about the type of ghosting they want. If they want an RM that
is only algebraic, then we shouldn't be adding ghosting functors
to the mesh

lindsayad committed:
Add CouplingFunctorCheck action

This action checks to see whether we should be adding
intra-element dof-coupling by inspecting the kernel and
boundary condition warehouses. If do need intra-dof coupling
then we add a default coupling functor and reinit the
libMesh::ImplicitSystem

Refs #13736

lindsayad committed:
Add test

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/371198/) on fdd2f14 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13775/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/371350/) on c9059a7 : invalidated by @permcody

lindsayad committed:
Don't add coupling functor for JFNK

lindsayad committed:
Make sure we have coupling matrices attached

If we add default coupling in the CouplingFunctorCheckAction
then we have to make sure that we have coupling matrices attached
to the coupling functor before calling DofMap::compute_sparsity
or else we will get the wrong sparsity pattern
